tools/data-handler/src/macros/BaseMacro.ts

```diff
@@ -0,0 +1,118 @@
+/**
+    Cyberismo
+    Copyright © Cyberismo Ltd and contributors 2024
+
+    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License version 3 as published by the Free Software Foundation.
+
+    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public
+    License along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { MacroGenerationContext, MacroMetadata } from './common.js';
+import { handleMacroError } from './index.js';
+
+abstract class BaseMacro {
+  /**
+   * Used for identifying the macro globally.
+   * This should be async safe.
+   */
+  private static globalCounter = 0;
+
+  private globalId: string;
+  private promiseList: {
+    localId: number;
+    promise: Promise<void>;
+    placeholder: string;
+    promiseResult: string | null;
+  }[] = [];
+
+  constructor(protected macroMetadata: MacroMetadata) {
+    // Set a unique global id for each instance of the macro, based on the macro name
+    this.globalId = `${this.macroMetadata.name}-${BaseMacro.globalCounter++}`;
+  }
+
+  protected abstract handleStatic(
+    context: MacroGenerationContext,
+    input: string,
+  ): Promise<string>;
+
+  protected abstract handleInject(
+    context: MacroGenerationContext,
+    input: string,
+  ): Promise<string>;
+
+  public get metadata() {
+    return this.macroMetadata;
+  }
+
+  /**
+   * Function responsible for starting the promise and storing it along with its localId.
+   */
+  public handleMacro = (context: MacroGenerationContext, input: string) => {
+    // Create a unique localId for each invocation
+    const localId = this.promiseList.length;
+    const placeholder = `${this.globalId}-${localId}`;
+
+    const func =
```

Comment:
Do not use abbreviations. Could this be ´functionToCall`? Or `boundFunction` or something similar?